content/en/blog/_posts/the-case-for-kubernetes-limits/index.md

```diff
@@ -0,0 +1,54 @@
+
+Now, with Black Friday and Cyber Monday approaching, businesses expect a surge in traffic. Good performance data/benchmarks of the past performance allow businesses to plan for some extra capacity. But is data about pods without limits reliable? With memory or CPU instant spikes handled by the extra resources, everything might look good according to past data. But once the pod bin-packing changes and the extra resources get more scarce, everything might start looking different – ranging from request latencies rising negligibly to requests slowly snowballing and causing pod OOM kills. While almost no one actually cares about the former, the latter is a serious issue that requires instant capacity increase.
+
+## Configuring the limits
+Not using limits takes a tradeoff – it opportunistically improves the performance if there are extra resources available, but lowers predictability of the performance, which might strike back in the future. There are a few approaches that can be used to increase the predictability again. Let’s pick two of them to analyze:
+
+- *Configure workload limits to be a fixed (and small) percentage more than the requests – let’s call it “fixed-fraction headroom".* This allows the use of some extra shared resources, but keeps the per-node overcommit bound and can be taken to guide worst-case estimates for the workload. Note that the bigger the limits percentage is, the bigger the variance in the performance that might happen across the workloads.
+- *Configure workloads with requests = limits.* From some point of view, this is equivalent to giving each pod its own tiny machine with constrained resources; the performance is fairly predictable. This also puts the pod into the Guaranteed QoS class, which makes it get evicted only after Best Effort and Burstable pods have been evicted by a node under resource pressure (see [here](https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/)).
+
+Some other cases might also be considered, but these are probably the two simplest ones to discuss.
+
```

Comment:
Which resources should people set requests and limits for? Just CPU and memory? Anything else that is sometimes worth defining?

Reply:
It's debatable, but I think CPU and memory is okay-ish here.

IIUC, the only resources that can be requested/limited by Kubernetes API are CPU and memory (and possibly hugepages, ephemeral-storage): https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-requests-and-limits-of-pod-and-container . I believe the former two cover most of the use-cases and is much easier to imagine than the latter two, so I'd rather focus on those.

As for defining more things -- I'd like things lean and not introduce additional strategies that might be hard to evaluate/do estimates on.